tenders/hvt/hvt_freebsd_x86_64.c

```diff
@@ -163,6 +163,11 @@ int hvt_vcpu_loop(struct hvt *hvt)
 {
     struct hvt_b *hvb = hvt->b;
     int ret;
+#if __FreeBSD_version >= 1400000
+    struct vm_exit vmexit;
+    hvb->vmrun.vm_exit = &vmexit;
+    struct vm_exit *vme = &vmexit;
+#endif
```

Comment:
I have no knowledge with FreeBSD, and fail to understand the how the code works (I just guess that the `vmrun.vm_exit` field was a `struct vm_exit` and now is a `struct vm_exit *`), so maybe I'm off-topic :)

I'd wrote both `#if` pragma at the same place, like:
```
#if __FreeBSD_version >= 1400000
    struct vm_exit vmexit;
    hvb->vmrun.vm_exit = &vmexit;
    struct vm_exit *vme = &vmexit;
#else
    struct vm_exit *vme = &hvb->vmrun.vm_exit;
#endif
```
What I miss is if `vme` should be re-set for each loop iteration (it was before 1400000, but not after 1400000).

Reply:
My reference is bhyverun.c -- in 13.2  https://github.com/freebsd/freebsd-src/blob/releng/13.2/usr.sbin/bhyve/bhyverun.c#L984-L988

and in 14.0 https://github.com/freebsd/freebsd-src/blob/releng/14.0/usr.sbin/bhyve/bhyverun.c#L961-L965

So, vmexit is stack-allocated - it is upshed into the `struct vm_run` (`hvb->vmrun.vm_exit = &vmexit`). Previously it was extracted after the ioctl VM_RUN was called. I guess for symmetry we could put the `struct vm_exit *vme` at the same place, but I would avoid to change the 13.x code path (that has been tested for years in production).

I'm fine with putting:
```C
#if __FreeBSD_version >= 1400000
    struct vm_exit vmexit;
    hvb->vmrun.vm_exit = &vmexit;
#endif
```

outside the loop, and:

```C
#if __FreeBSD_version >= 1400000
    struct vm_exit *vme = &vmexit;
#else
    struct vm_exit *vme = &hvb->vmrun.vm_exit;
#endif
```

inside the loop - if you prefer that code. :)

--

NB: Yes, in 13.x (and 12.x), the `struct vm_run` was:

```C
struct vm_run {
        int             cpuid;
        struct vm_exit  vm_exit;
};
```

while in a 14.x it has been revised to:

```C
struct vm_run {
        int             cpuid;
        cpuset_t        *cpuset;        /* CPU set storage */
        size_t          cpusetsize;
        struct vm_exit  *vm_exit;
};
```

Reply:
Oh, and thanks for your review :)

Reply:
That makes sense, thanks for the answer :)

Also regarding the new `vm_run` definition, shouldn't `cpuset` also be set to a non-NULL pointer (it probably shouldn't be, as `cpusetsize` is inited to 0). I was wondering if it could be better to enforce that with `memset(&hvb->vmrun, 0, sizeof(struct vm_run));` but `hvt` is `memset` to 0 in `hvt_init` so I think this is good :)